### DISCUSSION TITLE
Discussion: Restructure contrib/ folder

### DISCUSSION THREAD

**Discussion by Griatch** [evennia/evennia#2585] [Evennia Engine Development]
Looking for feedback on this, since it will be a backwards-incompatible change (you'll need to change your contrib imports). 

Up until now the `evennia/contrib/` folder has been growing organically, with everything in the same top-level structure. As per https://github.com/evennia/evennia/issues/1612, there is a suggestion of reorganizing it to be a little easier to navigate. 

This would be in addition to a more extensive readme/doc pages describing the contribs. And yes, this would mean people would have to change their imports - there's no way around that. It would be a one-time change though and easy to do. But it also means the change should be future proof from Evennia 1.0 and onward. For sub-dividing the contrib folder into categories, I don't think it's useful to divide by backend/user-facing (as the ticket suggest) because many contribs are fully systems that both contain backend systems as well as commands and user-facing parts in one package). 

Here's my current suggestion for refactoring the contrib/ structure:

- We make a policy that _all_ contribs must always be contained in their own folder. This may seem like overkill for small things like `slow_exit`, but it would also make it easy to store the unit-test together with each contrib rather than stuffing it all into `contrib/tests.py` like we do today. 
- To avoid too deep nesting, we can make things available in the top-level `contrib/__init__.py` as well as in `contrib/subfolder/contrib/__init__py` so that people can get contribs that way (normal Python packaging style). It may not be suitable for some contribs requiring optional third-party imports though.

For categorizing into subfolders, here's my current suggestion for folders under `evennia/contrib/` (the above suggestsions not applied):

- `grid/`   - contribs dealing with the game-world grid, maps and rooms
  - building_menu.py
  - extended_room.py
  - mapbuilder.py
  - simpledoor.py
  - slow_exit.py
  - wilderness.py
  - xyzgrid
- `rpg/`  - contribs related to rpg mechanics, rules and rolepay
  - chargen.py
  - dice.py
  - health_bar.py
  - gendersub.py
  - rplanguage.py
  - rpsystem.py
  - traits.py
- `game_systems/`  - other specific gameplay systems
  - barter.py
  - crafting/
  - clothing.py  
  - puzzles.py
  - ingame_python/
  - multidescer.py
  - mail.py
  - turnbattle/  
  - talking_npc.py
- `base_systems/`  - server base functionality changes
  - awsstorage/
  - color_markups.py
  - custom_gametime.py
  - email_login.py
  - menu_login.py
  - mux_comms_cmds.py
  - unixcommand.py
- `full_systems/`  - complete game engines
  - evscaperoom
- `utils/`  - various utilities for use anywhere 
  - fieldfill.py
  - random_string_generator.py
  - security/
  - tree_select.py
- `tutorials/` - tutorial resources
  - tutorial_examples/
  - tutorial_world/

<details>
<summary>[For reference, click here for alphabetical list of all contribs in develop branch]</summary>

- awsstorage/
- barter.py
- building_menu.py
- chargen.py
- clothing.py
- color_markups.py
- crafting/
- custom_gametime.py
- dice.py
- email_login.py
- evscaperoom
- extended_room.py
- fieldfill.py
- gendersub.py
- health_bar.py
- ingame_python/
- mail.py
- mapbuilder.py
- menu_login.py
- multidescer.py
- mux_comms_cmds.py
- puzzles.py
- random_string_generator.py
- rplanguage.py
- rpsystem.py
- security/
- simpledoor.py
- slow_exit.py
- talking_npc.py
- traits.py
- tree_select.py
- turnbattle/
- tutorial_examples
- tutorial_world
- unixcommand.py
- wilderness.py
- xyzgrid

</details>

**Comment by kovitikus** [comment]
I think reorganization is a great idea. I'm not personally affected by this change, as I currently don't use any contribs. If I were to approach one in the future, I would definitely prefer the better organized structure.

**Reply by Griatch** [reply]
@kovitikus :+1: Okay, thanks for the input!

**Reply by ChrisLR** [reply]
I agree with reorganizing, makes every contrib a bit easier to find.
+1

**Reply by jaborsh** [reply]
Thirded.

**Comment by Griatch** [comment]
Looks like people agree with this reorganization, so I'll go with this then.
